qualtran/bloqs/bookkeeping/partition.py

```diff
@@ -137,22 +138,24 @@ def _classical_partition(self, x: 'ClassicalValT') -> Dict[str, 'ClassicalValT']
             start += size
         return out_vals
 
-    def _classical_unpartition(self, **vals: 'ClassicalValT'):
-        out_vals = []
+    def _classical_unpartition_to_bits(self, **vals: 'ClassicalValT') -> NDArray[np.uint]:
+        out_vals: list[list[int]] = []
         for reg in self.regs:
             reg_val = vals[reg.name]
             if isinstance(reg_val, np.ndarray):
-                out_vals.append(ints_to_bits(reg_val.ravel(), reg.bitsize).ravel())
+                out_vals.extend(reg.dtype.to_bits(val) for val in reg_val.ravel())
```

Comment:
Why this change? Also, why this change only in the `_classical_unpartition` and not in `_classical_partition` ? Was the previous logic wrong for certain cases? Can you please add tests which pass with the new version and fail with the original one?

Reply:
I'm using this to convert the `cvs` to bits in `CtrlSpecAnd` which allows any dtype. But the current classical simulation protocol only supports integer values in `ClassicalValT`, so seems like it worked either way.

Reply:
Also `bits_to_ints` / `ints_to_bits` are faster than manually calling `to_bits` / `from_bits` on each element. (similar to https://github.com/quantumlib/Qualtran/issues/1072). So only changed the part I'm using for now.

Reply:
Sorry, I don't understand. What was the part which was failing earlier and now gets fixed? I presume even now, `CtrlSpecAnd` would support only `QUInt`, `QInt` and `QAny` types (or, types for which `bits_to_ints` works correctly) since the classical bits need to be partitioned correctly (right now, partitioning is still based on `bits_to_ints` but unpartitioning uses `dtype.to_bits()`) ?

Reply:
`CtrlSpecAnd` also works for `QFxp` as it's just bitwise equality

Reply:
I'm using this only for extracting the classical control bits in `CtrlSpecAnd` (i.e. `_flat_cvs`), so I never merge these bits back to get the constants. 

But this would break the classical simulator, but as the classical sim only supports ints now, we don't have any tests that use Fxp.

Reply:
I think for now, it's better to say that `CtrlSpecAnd` does not support `QFxp`, because in reality it does not. The decomposition would be incorrect because `Partition` and `Unpartition` doesn't support `QFxp`. So it's fine to fail early at `_flat_cvs` instead of having a wrong decomposition. 

The alternative would be to compute the `_flat_cvs` without invoking `Partition` so that the computation of `_flat_cvs` works for all data types and update `Partition` bloq for `_classical_unpartition` and `_classical_partition` whenever we are ready to do so. 

The current logic leaves the state of things in a complicatedly broken way where one direction works (maybe) but the other doesn't.

Reply:
I reverted these changes. For now CtrlSpecAnd only supports integer types (checked in post_init). 

Will add more support after the classical sim for QFxp is merged.